src/neo/Network/P2P/TaskManager.cs

```diff
@@ -274,56 +227,68 @@ public static Props Props(NeoSystem system)
                 }
-                if (!AssignSyncTask(failedTask)) return;
             }
 
-            int taskCounts = sessions.Values.Sum(p => p.IndexTasks.Count);
-            var highestBlockIndex = sessions.Values.Max(p => p.LastBlockIndex);
-            for (; taskCounts < MaxSyncTasksCount; taskCounts++)
+            Header last = Blockchain.Singleton.HeaderCache.Last;
+            uint currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
+            uint headerHeight = last?.Index ?? currentHeight;
+            UInt256 headerHash = last?.Hash ?? NativeContract.Ledger.CurrentHash(snapshot);
+            // When the number of AvailableTasks is no more than 0, no pending tasks of InventoryType.Block, it should process pending the tasks of headers
+            // If not HeaderTask pending to be processed it should ask for more Blocks
+            if ((!HasHeaderTask || globalTasks[HeaderTaskHash] < MaxConncurrentTasks) && headerHeight < session.LastBlockIndex && !Blockchain.Singleton.HeaderCache.Full)
             {
-                if (lastTaskIndex >= highestBlockIndex || lastTaskIndex >= currentHeight + InvPayload.MaxHashesCount) break;
-                if (!AssignSyncTask(++lastTaskIndex)) break;
+                session.Tasks[HeaderTaskHash] = DateTime.UtcNow;
+                IncrementGlobalTask(HeaderTaskHash);
+                session.RemoteNode.Tell(Message.Create(MessageCommand.GetHeaders, GetBlockByIndexPayload.Create(headerHeight)));
             }
-        }
-
-        private void SendPingMessage()
-        {
-            DataCache snapshot = Blockchain.Singleton.View;
-            uint currentHeight = NativeContract.Ledger.CurrentIndex(snapshot);
-            UInt256 currentHash = NativeContract.Ledger.CurrentHash(snapshot);
-            TrimmedBlock block = NativeContract.Ledger.GetTrimmedBlock(snapshot, currentHash);
-            foreach (KeyValuePair<IActorRef, TaskSession> item in sessions)
+            else if (currentHeight < session.LastBlockIndex)
             {
-                var node = item.Key;
-                var session = item.Value;
-
-                if (session.ExpireTime < TimeProvider.Current.UtcNow ||
-                     (block.Index >= session.LastBlockIndex &&
-                     TimeProvider.Current.UtcNow.ToTimestampMS() - PingCoolingOffPeriod >= block.Timestamp))
+                uint startHeight = currentHeight;
+                for (uint i = currentHeight + 1; i <= headerHeight; i++)
                 {
-                    if (session.InvTasks.Remove(MemPoolTaskHash))
-                    {
-                        node.Tell(Message.Create(MessageCommand.Mempool));
-                    }
-                    node.Tell(Message.Create(MessageCommand.Ping, PingPayload.Create(currentHeight)));
-                    session.ExpireTime = TimeProvider.Current.UtcNow.AddMilliseconds(PingCoolingOffPeriod);
+                    UInt256 nextHash = Blockchain.Singleton.HeaderCache[i].Hash;
```

Comment:
@erikzhang There is a problem in this line. If Blockchain is also persisting blocks and trying to remove items from header cache, a null pointer problem is likely to occur. Similiar problems might also occur in other places where header cache is called outside of Blockchain.

Reply:
I don't know why we need hash in indexed sync.

Reply:
@Qiao-Jin Please check it again

Reply:
We need the hash to check in global task I suppose

Reply:
But the block sync is based on indexes. Do we need this before?

Reply:
This is the original logic before #1397 I think:
![image](https://user-images.githubusercontent.com/43407364/106288790-71896680-6283-11eb-819f-6db9bd8c54b7.png)

Reply:
The code is to send `Ping`. But your code is to sync the blocks.

Reply:
Sending Ping is the newly added code. Original code is to sync the blocks..

Reply:
The old logic is to sync blocks by hashes. That's why we need to check the hashes in `globalTasks`. But now we sync blocks by indexes. I think the logic is different.

Reply:
How about a `HeaderTask` in `TaskManager`, a `Dictionary` which contains headers that have been requested? Upon `Blockchain` receiving new headers or Persisiting new blocks TaskManager can be informed to delete corresponding indexes from `HeaderTask`.

Reply:
I've fixed it.